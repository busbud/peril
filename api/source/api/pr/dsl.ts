import * as express from "express"
import winston from "../../logger"

import { GitHub } from "danger/distribution/platforms/GitHub"
import inline from "danger/distribution/runner/runners/inline"

import { getTemporaryAccessTokenForInstallation } from "../../api/github"
import { RunType } from "../../danger/danger_run"
import { executorForInstallation } from "../../danger/danger_runner"
import { getPerilPlatformForDSL } from "../../danger/peril_platform"
import { githubAPIForCommentable } from "../../github/events/utils/commenting"
import { PERIL_ORG_INSTALLATION_ID } from "../../globals"

export const prDSLRunner = async (req: express.Request, res: express.Response, _: express.NextFunction) => {
  winston.info("[router] -- Received OK")

  const query = req.query
  if (!query.owner) {
    return res.status(422).jsonp({ error: "No `owner` query param sent." })
  }

  if (!query.repo) {
    return res.status(422).jsonp({ error: "No `repo` query param sent." })
  }

  if (!query.number) {
    return res.status(422).jsonp({ error: "No `number` query param sent." })
  }

  // This has to be set for public usage.
  if (!PERIL_ORG_INSTALLATION_ID) {
    throw new Error("You can't support PR DSLs without setting up the PERIL_ORG_INSTALLATION_ID")
  }

  const token = await getTemporaryAccessTokenForInstallation(PERIL_ORG_INSTALLATION_ID)

  const ghDetails = {
    fullName: query.owner + "/" + query.repo,
    prID: query.number,
  }

  const githubAPI = githubAPIForCommentable(token, ghDetails.fullName, ghDetails.prID)

  const gh = GitHub(githubAPI)
  const platform = getPerilPlatformForDSL(RunType.pr, gh, {})

  const exec = await executorForInstallation(platform, inline, {})
  const dangerDSL = await exec.dslForDanger()

  // Remove this to reduce data
  if (dangerDSL.github) {
    dangerDSL.github.api = {} as any
  }

  // TODO: include Danger version number in JSON
  return res.status(400).jsonp({
    danger: dangerDSL,
    status: "OK",
  })
}
