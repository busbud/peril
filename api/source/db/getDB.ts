import { jsonDatabase } from "./json"
import { mongoDatabase } from "./mongo"

import { DatabaseAdaptor } from "."
import { RuntimeEnvironment } from "./runtimeEnv"

const isJest = typeof jest !== "undefined"

const hasJSONDef = !!process.env.DATABASE_JSON_FILE
const hasPerilAPIURL = !!process.env.PUBLIC_API_ROOT_URL

const hasLambdaEnv = !!process.env._HANDLER

/** There are three runtime environments for Peril, this says which one it is */
export const runtimeEnvironment = hasJSONDef
  ? RuntimeEnvironment.Standalone
  : hasPerilAPIURL
    ? RuntimeEnvironment.Peril
    : hasLambdaEnv
      ? RuntimeEnvironment.Runner
      : RuntimeEnvironment.Unknown

const getDatabaseForEnv = (env: any): DatabaseAdaptor | null => {
  if (env.DATABASE_JSON_FILE || isJest) {
    const json = jsonDatabase(env.DATABASE_JSON_FILE)
    json.setup()
    return json
  }

  if (env.MONGODB_URI) {
    if (!isJest) {
      mongoDatabase.setup()
    }
    return mongoDatabase
  }

  return null
}

let db: DatabaseAdaptor | null = null
/** Gets the Current DB for this runtime environment */
export const getDB = () => {
  if (!db) {
    db = getDatabaseForEnv(process.env)
  }

  if (!db) {
    throw new Error("No default DB was set up")
  }

  return db
}
