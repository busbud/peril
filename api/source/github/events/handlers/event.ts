import { DangerResults } from "danger/distribution/dsl/DangerResults"
import { GitHubAPI } from "danger/distribution/platforms/github/GitHubAPI"
import { octokitAPIForPeril } from "../../../danger/append_peril"
import { DangerRun, RunFeedback, RunType } from "../../../danger/danger_run"
import { runDangerForInstallation } from "../../../danger/danger_runner"
import { getGitHubFileContents } from "../../lib/github_helpers"
import { GitHubRunSettings } from "../github_runner"
import { githubAPIForCommentable } from "../utils/commenting"

export const runEventRun = async (
  eventName: string,
  runs: DangerRun[],
  settings: GitHubRunSettings,
  token: string,
  event: any
): Promise<DangerResults | null> => {
  if (!runs.length) {
    return null
  }

  const repoForDangerfile = runs[0].repoSlug || (event.repository && event.repository.full_name)
  if (!repoForDangerfile) {
    return null
  }
  // Can we actually provide feedback on this event?
  const supportsGithubCommentAPIs = runs[0].feedback === RunFeedback.commentable

  // Do we need an authenticated Danger GitHubAPI instance so we
  // can leave feedback on an issue?
  let githubAPI = null as GitHubAPI | null
  if (supportsGithubCommentAPIs && settings.commentableID && settings.repoName) {
    githubAPI = githubAPIForCommentable(token, settings.repoName, settings.commentableID)
  }

  const contents: string[] = []
  for (const run of runs) {
    const headDangerfile = await getGitHubFileContents(token, repoForDangerfile, run.dangerfilePath, run.branch)
    contents.push(headDangerfile)
  }

  const installationSettings = {
    iID: settings.installationID,
    settings: settings.installationSettings,
  }

  // Manually set the Octokit API object so that danger.github.api works
  const dangerDSL = {
    github: {
      ...event,
      api: await octokitAPIForPeril(settings.installationID, token),
    },
  }

  const results = await runDangerForInstallation(
    eventName,
    contents,
    runs.map(r => r.referenceString),
    githubAPI,
    RunType.import,
    installationSettings,
    { dsl: dangerDSL, webhook: event }
  )

  return results || null
}
