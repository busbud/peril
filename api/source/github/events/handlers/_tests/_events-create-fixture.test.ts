import { format } from "prettier"

jest.mock("../../../../db/getDB")
import { MockDB } from "../../../../db/__mocks__/getDB"
import { getDB } from "../../../../db/getDB"
const mockDB = getDB() as MockDB

jest.mock("../../../../api/github", () => ({
  getTemporaryAccessTokenForInstallation: () => Promise.resolve("12345"),
}))

jest.mock("../../../../runner/runFromSameHost")
import { runFromSameHost } from "../../../../runner/runFromSameHost"

jest.mock("../../../../github/lib/github_helpers", () => ({
  getGitHubFileContents: jest.fn(),
}))
import { getGitHubFileContents } from "../../../lib/github_helpers"
const mockGetGitHubFileContents: any = getGitHubFileContents

import { readFileSync, writeFileSync } from "fs"
import { resolve } from "path"
import { dangerRunForRules } from "../../../../danger/danger_run"

import { DangerDSLJSONType } from "danger/distribution/dsl/DangerDSL"
import { PerilRunnerBootstrapJSON } from "../../../../runner/triggerSandboxRun"
import { setupForRequest } from "../../github_runner"
import { runEventRun } from "../event"

const apiFixtures = resolve(__dirname, "../../_tests/fixtures")
const fixture = (file: string) => JSON.parse(readFileSync(resolve(apiFixtures, file), "utf8"))

it("passes the right args to the hyper functions", async () => {
  mockDB.getInstallation.mockReturnValue({ iID: "123", repos: {}, envVars: { hello: "world" } })

  const body = fixture("issue_comment_created.json")
  const req = { body, headers: { "X-GitHub-Delivery": "123" } } as any
  const settings = await setupForRequest(req, {})

  const dangerfileForRun = "warn(danger.github.api)"
  mockGetGitHubFileContents.mockImplementationOnce(() => Promise.resolve(dangerfileForRun))

  const runSettings = { issue_comment: "danger/peril-settings@testing/async_import.ts" }
  const run = dangerRunForRules("issue_comment", "created", runSettings, body)[0]

  await runEventRun("mockEvent", [run], settings, "12345", body)

  // Take the payload, remove the JWT and Github api and save a copy of the JSON into a fixture dir, then snapshot it
  const payload = (runFromSameHost as any).mock.calls[0][0] as PerilRunnerBootstrapJSON
  payload.perilSettings.perilJWT = "12345"
  payload.perilSettings.perilRunID = "[run-id]"

  // Ensure Github API is set before clearing it for the snapshot
  expect((payload.payload.dsl as DangerDSLJSONType).github!.api).toBeDefined()
  delete (payload.payload.dsl as DangerDSLJSONType).github!.api

  writeFileSync(
    __dirname + "/fixtures/PerilRunnerEventBootStrapExample.json",
    format(JSON.stringify(payload, null, "  "), { parser: "json" }),
    "utf8"
  )

  expect(payload).toMatchSnapshot()
})
