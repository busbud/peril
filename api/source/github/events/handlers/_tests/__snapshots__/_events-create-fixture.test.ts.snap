// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes the right args to the hyper functions 1`] = `
Object {
  "dslType": "run",
  "installation": Object {
    "iID": 23511,
    "settings": Object {},
  },
  "paths": Array [
    "danger/peril-settings@testing/async_import.ts",
  ],
  "payload": Object {
    "dsl": Object {
      "github": Object {
        "action": "created",
        "comment": Object {
          "body": "@mennenia Cool, sounds good",
          "created_at": "2017-06-30T17:20:55Z",
          "html_url": "https://github.com/artsy/emission/pull/631#issuecomment-312325112",
          "id": 312325112,
          "issue_url": "https://api.github.com/repos/artsy/emission/issues/631",
          "updated_at": "2017-06-30T17:20:55Z",
          "url": "https://api.github.com/repos/artsy/emission/issues/comments/312325112",
          "user": Object {
            "avatar_url": "https://avatars3.githubusercontent.com/u/296775?v=3",
            "events_url": "https://api.github.com/users/l2succes/events{/privacy}",
            "followers_url": "https://api.github.com/users/l2succes/followers",
            "following_url": "https://api.github.com/users/l2succes/following{/other_user}",
            "gists_url": "https://api.github.com/users/l2succes/gists{/gist_id}",
            "gravatar_id": "",
            "html_url": "https://github.com/l2succes",
            "id": 296775,
            "login": "l2succes",
            "organizations_url": "https://api.github.com/users/l2succes/orgs",
            "received_events_url": "https://api.github.com/users/l2succes/received_events",
            "repos_url": "https://api.github.com/users/l2succes/repos",
            "site_admin": false,
            "starred_url": "https://api.github.com/users/l2succes/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/l2succes/subscriptions",
            "type": "User",
            "url": "https://api.github.com/users/l2succes",
          },
        },
        "installation": Object {
          "id": 23511,
        },
        "issue": Object {
          "assignee": Object {
            "avatar_url": "https://avatars3.githubusercontent.com/u/373860?v=3",
            "events_url": "https://api.github.com/users/mennenia/events{/privacy}",
            "followers_url": "https://api.github.com/users/mennenia/followers",
            "following_url": "https://api.github.com/users/mennenia/following{/other_user}",
            "gists_url": "https://api.github.com/users/mennenia/gists{/gist_id}",
            "gravatar_id": "",
            "html_url": "https://github.com/mennenia",
            "id": 373860,
            "login": "mennenia",
            "organizations_url": "https://api.github.com/users/mennenia/orgs",
            "received_events_url": "https://api.github.com/users/mennenia/received_events",
            "repos_url": "https://api.github.com/users/mennenia/repos",
            "site_admin": false,
            "starred_url": "https://api.github.com/users/mennenia/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/mennenia/subscriptions",
            "type": "User",
            "url": "https://api.github.com/users/mennenia",
          },
          "assignees": Array [
            Object {
              "avatar_url": "https://avatars3.githubusercontent.com/u/373860?v=3",
              "events_url": "https://api.github.com/users/mennenia/events{/privacy}",
              "followers_url": "https://api.github.com/users/mennenia/followers",
              "following_url": "https://api.github.com/users/mennenia/following{/other_user}",
              "gists_url": "https://api.github.com/users/mennenia/gists{/gist_id}",
              "gravatar_id": "",
              "html_url": "https://github.com/mennenia",
              "id": 373860,
              "login": "mennenia",
              "organizations_url": "https://api.github.com/users/mennenia/orgs",
              "received_events_url": "https://api.github.com/users/mennenia/received_events",
              "repos_url": "https://api.github.com/users/mennenia/repos",
              "site_admin": false,
              "starred_url": "https://api.github.com/users/mennenia/starred{/owner}{/repo}",
              "subscriptions_url": "https://api.github.com/users/mennenia/subscriptions",
              "type": "User",
              "url": "https://api.github.com/users/mennenia",
            },
          ],
          "body": "This PR adds Pull to refresh support to the (Inbox) Active Bids + Messages view. It also fixes pagination.

Skip New Tests",
          "closed_at": null,
          "comments": 1,
          "comments_url": "https://api.github.com/repos/artsy/emission/issues/631/comments",
          "created_at": "2017-06-24T14:15:33Z",
          "events_url": "https://api.github.com/repos/artsy/emission/issues/631/events",
          "html_url": "https://github.com/artsy/emission/pull/631",
          "id": 238319730,
          "labels": Array [],
          "labels_url": "https://api.github.com/repos/artsy/emission/issues/631/labels{/name}",
          "locked": false,
          "milestone": null,
          "number": 631,
          "pull_request": Object {
            "diff_url": "https://github.com/artsy/emission/pull/631.diff",
            "html_url": "https://github.com/artsy/emission/pull/631",
            "patch_url": "https://github.com/artsy/emission/pull/631.patch",
            "url": "https://api.github.com/repos/artsy/emission/pulls/631",
          },
          "repository_url": "https://api.github.com/repos/artsy/emission",
          "state": "open",
          "title": "[MSG] Inbox pull to refresh & pagination",
          "updated_at": "2017-06-30T17:20:55Z",
          "url": "https://api.github.com/repos/artsy/emission/issues/631",
          "user": Object {
            "avatar_url": "https://avatars3.githubusercontent.com/u/296775?v=3",
            "events_url": "https://api.github.com/users/l2succes/events{/privacy}",
            "followers_url": "https://api.github.com/users/l2succes/followers",
            "following_url": "https://api.github.com/users/l2succes/following{/other_user}",
            "gists_url": "https://api.github.com/users/l2succes/gists{/gist_id}",
            "gravatar_id": "",
            "html_url": "https://github.com/l2succes",
            "id": 296775,
            "login": "l2succes",
            "organizations_url": "https://api.github.com/users/l2succes/orgs",
            "received_events_url": "https://api.github.com/users/l2succes/received_events",
            "repos_url": "https://api.github.com/users/l2succes/repos",
            "site_admin": false,
            "starred_url": "https://api.github.com/users/l2succes/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/l2succes/subscriptions",
            "type": "User",
            "url": "https://api.github.com/users/l2succes",
          },
        },
        "organization": Object {
          "avatar_url": "https://avatars0.githubusercontent.com/u/546231?v=3",
          "description": "Artsy.net Team",
          "events_url": "https://api.github.com/orgs/artsy/events",
          "hooks_url": "https://api.github.com/orgs/artsy/hooks",
          "id": 546231,
          "issues_url": "https://api.github.com/orgs/artsy/issues",
          "login": "artsy",
          "members_url": "https://api.github.com/orgs/artsy/members{/member}",
          "public_members_url": "https://api.github.com/orgs/artsy/public_members{/member}",
          "repos_url": "https://api.github.com/orgs/artsy/repos",
          "url": "https://api.github.com/orgs/artsy",
        },
        "repository": Object {
          "archive_url": "https://api.github.com/repos/artsy/emission/{archive_format}{/ref}",
          "assignees_url": "https://api.github.com/repos/artsy/emission/assignees{/user}",
          "blobs_url": "https://api.github.com/repos/artsy/emission/git/blobs{/sha}",
          "branches_url": "https://api.github.com/repos/artsy/emission/branches{/branch}",
          "clone_url": "https://github.com/artsy/emission.git",
          "collaborators_url": "https://api.github.com/repos/artsy/emission/collaborators{/collaborator}",
          "comments_url": "https://api.github.com/repos/artsy/emission/comments{/number}",
          "commits_url": "https://api.github.com/repos/artsy/emission/commits{/sha}",
          "compare_url": "https://api.github.com/repos/artsy/emission/compare/{base}...{head}",
          "contents_url": "https://api.github.com/repos/artsy/emission/contents/{+path}",
          "contributors_url": "https://api.github.com/repos/artsy/emission/contributors",
          "created_at": "2016-02-24T11:36:24Z",
          "default_branch": "master",
          "deployments_url": "https://api.github.com/repos/artsy/emission/deployments",
          "description": "React Native Components",
          "downloads_url": "https://api.github.com/repos/artsy/emission/downloads",
          "events_url": "https://api.github.com/repos/artsy/emission/events",
          "fork": false,
          "forks": 25,
          "forks_count": 25,
          "forks_url": "https://api.github.com/repos/artsy/emission/forks",
          "full_name": "artsy/emission",
          "git_commits_url": "https://api.github.com/repos/artsy/emission/git/commits{/sha}",
          "git_refs_url": "https://api.github.com/repos/artsy/emission/git/refs{/sha}",
          "git_tags_url": "https://api.github.com/repos/artsy/emission/git/tags{/sha}",
          "git_url": "git://github.com/artsy/emission.git",
          "has_downloads": true,
          "has_issues": true,
          "has_pages": false,
          "has_projects": false,
          "has_wiki": false,
          "homepage": null,
          "hooks_url": "https://api.github.com/repos/artsy/emission/hooks",
          "html_url": "https://github.com/artsy/emission",
          "id": 52437732,
          "issue_comment_url": "https://api.github.com/repos/artsy/emission/issues/comments{/number}",
          "issue_events_url": "https://api.github.com/repos/artsy/emission/issues/events{/number}",
          "issues_url": "https://api.github.com/repos/artsy/emission/issues{/number}",
          "keys_url": "https://api.github.com/repos/artsy/emission/keys{/key_id}",
          "labels_url": "https://api.github.com/repos/artsy/emission/labels{/name}",
          "language": "TypeScript",
          "languages_url": "https://api.github.com/repos/artsy/emission/languages",
          "merges_url": "https://api.github.com/repos/artsy/emission/merges",
          "milestones_url": "https://api.github.com/repos/artsy/emission/milestones{/number}",
          "mirror_url": null,
          "name": "emission",
          "notifications_url": "https://api.github.com/repos/artsy/emission/notifications{?since,all,participating}",
          "open_issues": 70,
          "open_issues_count": 70,
          "owner": Object {
            "avatar_url": "https://avatars0.githubusercontent.com/u/546231?v=3",
            "events_url": "https://api.github.com/users/artsy/events{/privacy}",
            "followers_url": "https://api.github.com/users/artsy/followers",
            "following_url": "https://api.github.com/users/artsy/following{/other_user}",
            "gists_url": "https://api.github.com/users/artsy/gists{/gist_id}",
            "gravatar_id": "",
            "html_url": "https://github.com/artsy",
            "id": 546231,
            "login": "artsy",
            "organizations_url": "https://api.github.com/users/artsy/orgs",
            "received_events_url": "https://api.github.com/users/artsy/received_events",
            "repos_url": "https://api.github.com/users/artsy/repos",
            "site_admin": false,
            "starred_url": "https://api.github.com/users/artsy/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/artsy/subscriptions",
            "type": "Organization",
            "url": "https://api.github.com/users/artsy",
          },
          "private": false,
          "pulls_url": "https://api.github.com/repos/artsy/emission/pulls{/number}",
          "pushed_at": "2017-06-30T17:20:18Z",
          "releases_url": "https://api.github.com/repos/artsy/emission/releases{/id}",
          "size": 10179,
          "ssh_url": "git@github.com:artsy/emission.git",
          "stargazers_count": 223,
          "stargazers_url": "https://api.github.com/repos/artsy/emission/stargazers",
          "statuses_url": "https://api.github.com/repos/artsy/emission/statuses/{sha}",
          "subscribers_url": "https://api.github.com/repos/artsy/emission/subscribers",
          "subscription_url": "https://api.github.com/repos/artsy/emission/subscription",
          "svn_url": "https://github.com/artsy/emission",
          "tags_url": "https://api.github.com/repos/artsy/emission/tags",
          "teams_url": "https://api.github.com/repos/artsy/emission/teams",
          "trees_url": "https://api.github.com/repos/artsy/emission/git/trees{/sha}",
          "updated_at": "2017-06-30T12:52:37Z",
          "url": "https://api.github.com/repos/artsy/emission",
          "watchers": 223,
          "watchers_count": 223,
        },
        "sender": Object {
          "avatar_url": "https://avatars3.githubusercontent.com/u/296775?v=3",
          "events_url": "https://api.github.com/users/l2succes/events{/privacy}",
          "followers_url": "https://api.github.com/users/l2succes/followers",
          "following_url": "https://api.github.com/users/l2succes/following{/other_user}",
          "gists_url": "https://api.github.com/users/l2succes/gists{/gist_id}",
          "gravatar_id": "",
          "html_url": "https://github.com/l2succes",
          "id": 296775,
          "login": "l2succes",
          "organizations_url": "https://api.github.com/users/l2succes/orgs",
          "received_events_url": "https://api.github.com/users/l2succes/received_events",
          "repos_url": "https://api.github.com/users/l2succes/repos",
          "site_admin": false,
          "starred_url": "https://api.github.com/users/l2succes/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/l2succes/subscriptions",
          "type": "User",
          "url": "https://api.github.com/users/l2succes",
        },
      },
      "settings": Object {
        "cliArgs": Object {},
        "github": Object {
          "accessToken": "12345",
          "additionalHeaders": Object {
            "Accept": "application/vnd.github.machine-man-preview+json",
          },
          "baseURL": undefined,
        },
      },
    },
    "webhook": Object {
      "action": "created",
      "comment": Object {
        "body": "@mennenia Cool, sounds good",
        "created_at": "2017-06-30T17:20:55Z",
        "html_url": "https://github.com/artsy/emission/pull/631#issuecomment-312325112",
        "id": 312325112,
        "issue_url": "https://api.github.com/repos/artsy/emission/issues/631",
        "updated_at": "2017-06-30T17:20:55Z",
        "url": "https://api.github.com/repos/artsy/emission/issues/comments/312325112",
        "user": Object {
          "avatar_url": "https://avatars3.githubusercontent.com/u/296775?v=3",
          "events_url": "https://api.github.com/users/l2succes/events{/privacy}",
          "followers_url": "https://api.github.com/users/l2succes/followers",
          "following_url": "https://api.github.com/users/l2succes/following{/other_user}",
          "gists_url": "https://api.github.com/users/l2succes/gists{/gist_id}",
          "gravatar_id": "",
          "html_url": "https://github.com/l2succes",
          "id": 296775,
          "login": "l2succes",
          "organizations_url": "https://api.github.com/users/l2succes/orgs",
          "received_events_url": "https://api.github.com/users/l2succes/received_events",
          "repos_url": "https://api.github.com/users/l2succes/repos",
          "site_admin": false,
          "starred_url": "https://api.github.com/users/l2succes/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/l2succes/subscriptions",
          "type": "User",
          "url": "https://api.github.com/users/l2succes",
        },
      },
      "installation": Object {
        "id": 23511,
      },
      "issue": Object {
        "assignee": Object {
          "avatar_url": "https://avatars3.githubusercontent.com/u/373860?v=3",
          "events_url": "https://api.github.com/users/mennenia/events{/privacy}",
          "followers_url": "https://api.github.com/users/mennenia/followers",
          "following_url": "https://api.github.com/users/mennenia/following{/other_user}",
          "gists_url": "https://api.github.com/users/mennenia/gists{/gist_id}",
          "gravatar_id": "",
          "html_url": "https://github.com/mennenia",
          "id": 373860,
          "login": "mennenia",
          "organizations_url": "https://api.github.com/users/mennenia/orgs",
          "received_events_url": "https://api.github.com/users/mennenia/received_events",
          "repos_url": "https://api.github.com/users/mennenia/repos",
          "site_admin": false,
          "starred_url": "https://api.github.com/users/mennenia/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/mennenia/subscriptions",
          "type": "User",
          "url": "https://api.github.com/users/mennenia",
        },
        "assignees": Array [
          Object {
            "avatar_url": "https://avatars3.githubusercontent.com/u/373860?v=3",
            "events_url": "https://api.github.com/users/mennenia/events{/privacy}",
            "followers_url": "https://api.github.com/users/mennenia/followers",
            "following_url": "https://api.github.com/users/mennenia/following{/other_user}",
            "gists_url": "https://api.github.com/users/mennenia/gists{/gist_id}",
            "gravatar_id": "",
            "html_url": "https://github.com/mennenia",
            "id": 373860,
            "login": "mennenia",
            "organizations_url": "https://api.github.com/users/mennenia/orgs",
            "received_events_url": "https://api.github.com/users/mennenia/received_events",
            "repos_url": "https://api.github.com/users/mennenia/repos",
            "site_admin": false,
            "starred_url": "https://api.github.com/users/mennenia/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/mennenia/subscriptions",
            "type": "User",
            "url": "https://api.github.com/users/mennenia",
          },
        ],
        "body": "This PR adds Pull to refresh support to the (Inbox) Active Bids + Messages view. It also fixes pagination.

Skip New Tests",
        "closed_at": null,
        "comments": 1,
        "comments_url": "https://api.github.com/repos/artsy/emission/issues/631/comments",
        "created_at": "2017-06-24T14:15:33Z",
        "events_url": "https://api.github.com/repos/artsy/emission/issues/631/events",
        "html_url": "https://github.com/artsy/emission/pull/631",
        "id": 238319730,
        "labels": Array [],
        "labels_url": "https://api.github.com/repos/artsy/emission/issues/631/labels{/name}",
        "locked": false,
        "milestone": null,
        "number": 631,
        "pull_request": Object {
          "diff_url": "https://github.com/artsy/emission/pull/631.diff",
          "html_url": "https://github.com/artsy/emission/pull/631",
          "patch_url": "https://github.com/artsy/emission/pull/631.patch",
          "url": "https://api.github.com/repos/artsy/emission/pulls/631",
        },
        "repository_url": "https://api.github.com/repos/artsy/emission",
        "state": "open",
        "title": "[MSG] Inbox pull to refresh & pagination",
        "updated_at": "2017-06-30T17:20:55Z",
        "url": "https://api.github.com/repos/artsy/emission/issues/631",
        "user": Object {
          "avatar_url": "https://avatars3.githubusercontent.com/u/296775?v=3",
          "events_url": "https://api.github.com/users/l2succes/events{/privacy}",
          "followers_url": "https://api.github.com/users/l2succes/followers",
          "following_url": "https://api.github.com/users/l2succes/following{/other_user}",
          "gists_url": "https://api.github.com/users/l2succes/gists{/gist_id}",
          "gravatar_id": "",
          "html_url": "https://github.com/l2succes",
          "id": 296775,
          "login": "l2succes",
          "organizations_url": "https://api.github.com/users/l2succes/orgs",
          "received_events_url": "https://api.github.com/users/l2succes/received_events",
          "repos_url": "https://api.github.com/users/l2succes/repos",
          "site_admin": false,
          "starred_url": "https://api.github.com/users/l2succes/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/l2succes/subscriptions",
          "type": "User",
          "url": "https://api.github.com/users/l2succes",
        },
      },
      "organization": Object {
        "avatar_url": "https://avatars0.githubusercontent.com/u/546231?v=3",
        "description": "Artsy.net Team",
        "events_url": "https://api.github.com/orgs/artsy/events",
        "hooks_url": "https://api.github.com/orgs/artsy/hooks",
        "id": 546231,
        "issues_url": "https://api.github.com/orgs/artsy/issues",
        "login": "artsy",
        "members_url": "https://api.github.com/orgs/artsy/members{/member}",
        "public_members_url": "https://api.github.com/orgs/artsy/public_members{/member}",
        "repos_url": "https://api.github.com/orgs/artsy/repos",
        "url": "https://api.github.com/orgs/artsy",
      },
      "repository": Object {
        "archive_url": "https://api.github.com/repos/artsy/emission/{archive_format}{/ref}",
        "assignees_url": "https://api.github.com/repos/artsy/emission/assignees{/user}",
        "blobs_url": "https://api.github.com/repos/artsy/emission/git/blobs{/sha}",
        "branches_url": "https://api.github.com/repos/artsy/emission/branches{/branch}",
        "clone_url": "https://github.com/artsy/emission.git",
        "collaborators_url": "https://api.github.com/repos/artsy/emission/collaborators{/collaborator}",
        "comments_url": "https://api.github.com/repos/artsy/emission/comments{/number}",
        "commits_url": "https://api.github.com/repos/artsy/emission/commits{/sha}",
        "compare_url": "https://api.github.com/repos/artsy/emission/compare/{base}...{head}",
        "contents_url": "https://api.github.com/repos/artsy/emission/contents/{+path}",
        "contributors_url": "https://api.github.com/repos/artsy/emission/contributors",
        "created_at": "2016-02-24T11:36:24Z",
        "default_branch": "master",
        "deployments_url": "https://api.github.com/repos/artsy/emission/deployments",
        "description": "React Native Components",
        "downloads_url": "https://api.github.com/repos/artsy/emission/downloads",
        "events_url": "https://api.github.com/repos/artsy/emission/events",
        "fork": false,
        "forks": 25,
        "forks_count": 25,
        "forks_url": "https://api.github.com/repos/artsy/emission/forks",
        "full_name": "artsy/emission",
        "git_commits_url": "https://api.github.com/repos/artsy/emission/git/commits{/sha}",
        "git_refs_url": "https://api.github.com/repos/artsy/emission/git/refs{/sha}",
        "git_tags_url": "https://api.github.com/repos/artsy/emission/git/tags{/sha}",
        "git_url": "git://github.com/artsy/emission.git",
        "has_downloads": true,
        "has_issues": true,
        "has_pages": false,
        "has_projects": false,
        "has_wiki": false,
        "homepage": null,
        "hooks_url": "https://api.github.com/repos/artsy/emission/hooks",
        "html_url": "https://github.com/artsy/emission",
        "id": 52437732,
        "issue_comment_url": "https://api.github.com/repos/artsy/emission/issues/comments{/number}",
        "issue_events_url": "https://api.github.com/repos/artsy/emission/issues/events{/number}",
        "issues_url": "https://api.github.com/repos/artsy/emission/issues{/number}",
        "keys_url": "https://api.github.com/repos/artsy/emission/keys{/key_id}",
        "labels_url": "https://api.github.com/repos/artsy/emission/labels{/name}",
        "language": "TypeScript",
        "languages_url": "https://api.github.com/repos/artsy/emission/languages",
        "merges_url": "https://api.github.com/repos/artsy/emission/merges",
        "milestones_url": "https://api.github.com/repos/artsy/emission/milestones{/number}",
        "mirror_url": null,
        "name": "emission",
        "notifications_url": "https://api.github.com/repos/artsy/emission/notifications{?since,all,participating}",
        "open_issues": 70,
        "open_issues_count": 70,
        "owner": Object {
          "avatar_url": "https://avatars0.githubusercontent.com/u/546231?v=3",
          "events_url": "https://api.github.com/users/artsy/events{/privacy}",
          "followers_url": "https://api.github.com/users/artsy/followers",
          "following_url": "https://api.github.com/users/artsy/following{/other_user}",
          "gists_url": "https://api.github.com/users/artsy/gists{/gist_id}",
          "gravatar_id": "",
          "html_url": "https://github.com/artsy",
          "id": 546231,
          "login": "artsy",
          "organizations_url": "https://api.github.com/users/artsy/orgs",
          "received_events_url": "https://api.github.com/users/artsy/received_events",
          "repos_url": "https://api.github.com/users/artsy/repos",
          "site_admin": false,
          "starred_url": "https://api.github.com/users/artsy/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/artsy/subscriptions",
          "type": "Organization",
          "url": "https://api.github.com/users/artsy",
        },
        "private": false,
        "pulls_url": "https://api.github.com/repos/artsy/emission/pulls{/number}",
        "pushed_at": "2017-06-30T17:20:18Z",
        "releases_url": "https://api.github.com/repos/artsy/emission/releases{/id}",
        "size": 10179,
        "ssh_url": "git@github.com:artsy/emission.git",
        "stargazers_count": 223,
        "stargazers_url": "https://api.github.com/repos/artsy/emission/stargazers",
        "statuses_url": "https://api.github.com/repos/artsy/emission/statuses/{sha}",
        "subscribers_url": "https://api.github.com/repos/artsy/emission/subscribers",
        "subscription_url": "https://api.github.com/repos/artsy/emission/subscription",
        "svn_url": "https://github.com/artsy/emission",
        "tags_url": "https://api.github.com/repos/artsy/emission/tags",
        "teams_url": "https://api.github.com/repos/artsy/emission/teams",
        "trees_url": "https://api.github.com/repos/artsy/emission/git/trees{/sha}",
        "updated_at": "2017-06-30T12:52:37Z",
        "url": "https://api.github.com/repos/artsy/emission",
        "watchers": 223,
        "watchers_count": 223,
      },
      "sender": Object {
        "avatar_url": "https://avatars3.githubusercontent.com/u/296775?v=3",
        "events_url": "https://api.github.com/users/l2succes/events{/privacy}",
        "followers_url": "https://api.github.com/users/l2succes/followers",
        "following_url": "https://api.github.com/users/l2succes/following{/other_user}",
        "gists_url": "https://api.github.com/users/l2succes/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/l2succes",
        "id": 296775,
        "login": "l2succes",
        "organizations_url": "https://api.github.com/users/l2succes/orgs",
        "received_events_url": "https://api.github.com/users/l2succes/received_events",
        "repos_url": "https://api.github.com/users/l2succes/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/l2succes/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/l2succes/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/l2succes",
      },
    },
  },
  "perilSettings": Object {
    "envVars": Object {
      "hello": "world",
    },
    "event": "mockEvent",
    "perilAPIRoot": undefined,
    "perilJWT": "12345",
    "perilRunID": "[run-id]",
  },
}
`;
